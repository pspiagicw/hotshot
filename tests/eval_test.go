package tests

import (
	"reflect"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/pspiagicw/hotshot/eval"
	"github.com/pspiagicw/hotshot/lexer"
	"github.com/pspiagicw/hotshot/object"
	"github.com/pspiagicw/hotshot/parser"
)

func TestEval(t *testing.T) {

	// input := "()"
	tt := map[string]object.Object{
		"()":                           createNull(),
		"; this is a simple comment ;": createNull(),

		"1":               createInt(1),
		"23234":           createInt(23234),
		"(+ 1 2)":         createInt(3),
		"(+ 1 (+ 1 2) 2)": createInt(6),
		"(- 2 1 )":        createInt(1),
		"(* 5 6)":         createInt(30),
		"(+ 5 (- 4 5))":   createInt(4),
		"(+ -5 (* 4 5))":  createInt(15),
		// "(% 4 3)":             createInt(1),
		`($ number 5) number`: createInt(5),

		`"someString"`:            createString("someString"),
		`($ name "hotshot") name`: createString("hotshot"),

		`(= 1 1)`:           createBool(true),
		`(= 1 2)`:           createBool(false),
		`(= "some" "some")`: createBool(true),
		`(= "some" "else")`: createBool(false),

		`(= true true)`:   createBool(true),
		`(= false true)`:  createBool(false),
		`(= false false)`: createBool(true),

		`(> 1 1)`: createBool(false),
		`(< 1 1)`: createBool(false),
		`(< 2 1)`: createBool(false),
		`(> 2 1)`: createBool(true),

		`(not true)`:        createBool(false),
		`(not false)`:       createBool(true),
		`(and false true)`:  createBool(false),
		`(and true true)`:   createBool(true),
		`(and false false)`: createBool(false),
		`(or true false)`:   createBool(true),
		`(or false false)`:  createBool(false),
		`(or true true)`:    createBool(true),

		`(if true 5 2)`:    createInt(5),
		`(if false 5 2)`:   createInt(2),
		`(if (= 1 2) 5 2)`: createInt(2),
		`(if (= 1 1) 7 2)`: createInt(7),

		`(fn hello () "Hello, World")`:         createNull(),
		`(fn hello () "Hello, World") (hello)`: createString("Hello, World"),
		`(fn add (x y) (+ x y)) (add 2 1)`:     createInt(3),
	}

	for input, expectedResult := range tt {
		t.Run(input, func(t *testing.T) {
			checkResult(t, input, expectedResult)
		})
	}

}
func createNull() *object.Null {
	return &object.Null{}
}
func createString(val string) *object.String {
	return &object.String{
		Value: val,
	}
}
func createBool(val bool) *object.Boolean {
	return &object.Boolean{
		Value: val,
	}
}
func createInt(val int) *object.Integer {
	return &object.Integer{
		Value: val,
	}
}
func checkResult(t *testing.T, input string, expected object.Object) {

	t.Helper()

	lexer := lexer.NewLexer(input)
	parser := parser.NewParser(lexer)

	ast := parser.Parse()
	if len(parser.Errors()) != 0 {
		t.Fatalf("Errors while parsing")
	}

	snaps.MatchSnapshot(t, ast)

	// TODO: Last statement is always nil!
	ast.Statements = ast.Statements[:len(ast.Statements)-1]

	env := object.NewEnvironment()

	result := eval.Eval(ast, env)

	equalResult(t, result, expected)

}

func equalResult(t *testing.T, result, expected object.Object) {

	t.Helper()

	if result == nil {
		t.Fatalf("Evaluted result is nil!")
	}

	if result.Type() == object.NULL_OBJ && expected.Type() != object.NULL_OBJ {
		t.Errorf("Evaluated result not equal! got %v, expected Null", result)
	}

	if result.Type() != expected.Type() {
		t.Errorf("Evaluated result type not equal! got %v, expected %v", result.Type(), expected.Type())

	}

	if result.Type() != object.NULL_OBJ {
		if !reflect.DeepEqual(result, expected) {
			t.Errorf("Evaluated result not equal! got %v, expected %v", result, expected)
		}
	}
}
