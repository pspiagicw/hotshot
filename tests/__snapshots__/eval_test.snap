
[TestEval/(-_2_1_) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"MINUS", TokenValue:"-"},
            Args: {
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(+_-5_(*_4_5)) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
            Args: {
                &ast.IntStatement{Value:-5},
                &ast.CallStatement{
                    Op:   &token.Token{TokenType:"MULTIPLY", TokenValue:"*"},
                    Args: {
                        &ast.IntStatement{Value:4},
                        &ast.IntStatement{Value:5},
                    },
                },
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_"some"_"some") - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.StringStatement{Value:"some"},
                &ast.StringStatement{Value:"some"},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(if_(=_1_2)_5_2) - 1]
&ast.Program{
    Statements: {
        &ast.IfStatement{
            Condition: &ast.CallStatement{
                Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
                Args: {
                    &ast.IntStatement{Value:1},
                    &ast.IntStatement{Value:2},
                },
            },
            Body: &ast.IntStatement{Value:5},
            Else: &ast.IntStatement{Value:2},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/1 - 1]
&ast.Program{
    Statements: {
        &ast.IntStatement{Value:1},
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(<_2_1) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"LESSTHAN", TokenValue:"<"},
            Args: {
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(and_true_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"and"},
            Args: {
                &ast.BoolStatement{Value:true},
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/() - 1]
&ast.Program{
    Statements: {
        &ast.EmptyStatement{},
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(+_5_(-_4_5)) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
            Args: {
                &ast.IntStatement{Value:5},
                &ast.CallStatement{
                    Op:   &token.Token{TokenType:"MINUS", TokenValue:"-"},
                    Args: {
                        &ast.IntStatement{Value:4},
                        &ast.IntStatement{Value:5},
                    },
                },
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_"some"_"else") - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.StringStatement{Value:"some"},
                &ast.StringStatement{Value:"else"},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_true_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.BoolStatement{Value:true},
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(>_2_1) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"GREATERTHAN", TokenValue:">"},
            Args: {
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(not_false) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"not"},
            Args: {
                &ast.BoolStatement{},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(or_true_false) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"or"},
            Args: {
                &ast.BoolStatement{Value:true},
                &ast.BoolStatement{},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/;_this_is_a_simple_comment_; - 1]
&ast.Program{
    Statements: {
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(<_1_1) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"LESSTHAN", TokenValue:"<"},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(or_false_false) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"or"},
            Args: {
                &ast.BoolStatement{},
                &ast.BoolStatement{},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(if_false_5_2) - 1]
&ast.Program{
    Statements: {
        &ast.IfStatement{
            Condition: &ast.BoolStatement{},
            Body:      &ast.IntStatement{Value:5},
            Else:      &ast.IntStatement{Value:2},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_1_2) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.IntStatement{Value:2},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_1_1) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_false_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.BoolStatement{},
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(not_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"not"},
            Args: {
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(or_true_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"or"},
            Args: {
                &ast.BoolStatement{Value:true},
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/"someString" - 1]
&ast.Program{
    Statements: {
        &ast.StringStatement{Value:"someString"},
        &ast.EmptyStatement{},
    },
}
---

[TestEval/($_name_"hotshot")_name - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"name"},
            Value: &ast.StringStatement{Value:"hotshot"},
        },
        &ast.IdentStatement{
            Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(+_1_2) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.IntStatement{Value:2},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(>_1_1) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"GREATERTHAN", TokenValue:">"},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(and_false_true) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"and"},
            Args: {
                &ast.BoolStatement{},
                &ast.BoolStatement{Value:true},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(and_false_false) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"and"},
            Args: {
                &ast.BoolStatement{},
                &ast.BoolStatement{},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(if_(=_1_1)_7_2) - 1]
&ast.Program{
    Statements: {
        &ast.IfStatement{
            Condition: &ast.CallStatement{
                Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
                Args: {
                    &ast.IntStatement{Value:1},
                    &ast.IntStatement{Value:1},
                },
            },
            Body: &ast.IntStatement{Value:7},
            Else: &ast.IntStatement{Value:2},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/($_number_5)_number - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"number"},
            Value: &ast.IntStatement{Value:5},
        },
        &ast.IdentStatement{
            Value: &token.Token{TokenType:"IDENT", TokenValue:"number"},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(+_1_(+_1_2)_2) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
            Args: {
                &ast.IntStatement{Value:1},
                &ast.CallStatement{
                    Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
                    Args: {
                        &ast.IntStatement{Value:1},
                        &ast.IntStatement{Value:2},
                    },
                },
                &ast.IntStatement{Value:2},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(*_5_6) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"MULTIPLY", TokenValue:"*"},
            Args: {
                &ast.IntStatement{Value:5},
                &ast.IntStatement{Value:6},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(=_false_false) - 1]
&ast.Program{
    Statements: {
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"EQ", TokenValue:"="},
            Args: {
                &ast.BoolStatement{},
                &ast.BoolStatement{},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(if_true_5_2) - 1]
&ast.Program{
    Statements: {
        &ast.IfStatement{
            Condition: &ast.BoolStatement{Value:true},
            Body:      &ast.IntStatement{Value:5},
            Else:      &ast.IntStatement{Value:2},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/23234 - 1]
&ast.Program{
    Statements: {
        &ast.IntStatement{Value:23234},
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_hello_()_"Hello,_World") - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
            },
            Body: &ast.StringStatement{Value:"Hello, World"},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_hello_()_"Hello,_World")_(hello) - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
            },
            Body: &ast.StringStatement{Value:"Hello, World"},
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: nil,
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_add_(x_y)_(+_x_y))_(add_2_1) - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"add"},
            Args: {
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                },
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                },
            },
            Body: &ast.CallStatement{
                Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
                Args: {
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                    },
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                    },
                },
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"add"},
            Args: {
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_hello_(name)_(?_"Hello"_name))_(hello_"pspiagicw") - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
                },
            },
            Body: &ast.CallStatement{
                Op:   &token.Token{TokenType:"QUESTION", TokenValue:"?"},
                Args: {
                    &ast.StringStatement{Value:"Hello"},
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
                    },
                },
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
                &ast.StringStatement{Value:"pspiagicw"},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/($_a_(lambda_()_4))_(a) - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"a"},
            Value: &ast.LambdaStatement{
                Args: {
                },
                Body: &ast.IntStatement{Value:4},
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"a"},
            Args: nil,
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_arithmetic(op_x_y)_(op_x_y))_________(arithmetic_(lambda_(x_y)_(+_x_y))_2_1) - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"arithmetic"},
            Args: {
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"op"},
                },
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                },
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                },
            },
            Body: &ast.CallStatement{
                Op:   &token.Token{TokenType:"IDENT", TokenValue:"op"},
                Args: {
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                    },
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                    },
                },
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"arithmetic"},
            Args: {
                &ast.LambdaStatement{
                    Args: {
                        &ast.IdentStatement{
                            Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                        },
                        &ast.IdentStatement{
                            Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                        },
                    },
                    Body: &ast.CallStatement{
                        Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
                        Args: {
                            &ast.IdentStatement{
                                Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                            },
                            &ast.IdentStatement{
                                Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                            },
                        },
                    },
                },
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(let_name_"hotshot")_name - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"name"},
            Value: &ast.StringStatement{Value:"hotshot"},
        },
        &ast.IdentStatement{
            Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(let_a_(lambda_()_4))_(a) - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"a"},
            Value: &ast.LambdaStatement{
                Args: {
                },
                Body: &ast.IntStatement{Value:4},
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"a"},
            Args: nil,
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(let_number_5)_number - 1]
&ast.Program{
    Statements: {
        &ast.AssignmentStatement{
            Name:  &token.Token{TokenType:"IDENT", TokenValue:"number"},
            Value: &ast.IntStatement{Value:5},
        },
        &ast.IdentStatement{
            Value: &token.Token{TokenType:"IDENT", TokenValue:"number"},
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_hello_(name)_(echo_"Hello"_name))_(hello_"pspiagicw") - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
                },
            },
            Body: &ast.CallStatement{
                Op:   &token.Token{TokenType:"IDENT", TokenValue:"echo"},
                Args: {
                    &ast.StringStatement{Value:"Hello"},
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"name"},
                    },
                },
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"hello"},
            Args: {
                &ast.StringStatement{Value:"pspiagicw"},
            },
        },
        &ast.EmptyStatement{},
    },
}
---

[TestEval/(fn_arithmetic(op_x_y)_(op_x_y))_____________(arithmetic_(lambda_(x_y)_(+_x_y))_2_1) - 1]
&ast.Program{
    Statements: {
        &ast.FunctionStatement{
            Name: &token.Token{TokenType:"IDENT", TokenValue:"arithmetic"},
            Args: {
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"op"},
                },
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                },
                &ast.IdentStatement{
                    Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                },
            },
            Body: &ast.CallStatement{
                Op:   &token.Token{TokenType:"IDENT", TokenValue:"op"},
                Args: {
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                    },
                    &ast.IdentStatement{
                        Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                    },
                },
            },
        },
        &ast.CallStatement{
            Op:   &token.Token{TokenType:"IDENT", TokenValue:"arithmetic"},
            Args: {
                &ast.LambdaStatement{
                    Args: {
                        &ast.IdentStatement{
                            Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                        },
                        &ast.IdentStatement{
                            Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                        },
                    },
                    Body: &ast.CallStatement{
                        Op:   &token.Token{TokenType:"PLUS", TokenValue:"+"},
                        Args: {
                            &ast.IdentStatement{
                                Value: &token.Token{TokenType:"IDENT", TokenValue:"x"},
                            },
                            &ast.IdentStatement{
                                Value: &token.Token{TokenType:"IDENT", TokenValue:"y"},
                            },
                        },
                    },
                },
                &ast.IntStatement{Value:2},
                &ast.IntStatement{Value:1},
            },
        },
        &ast.EmptyStatement{},
    },
}
---
